[WiP] Make sure to always only have valid collection names

tmadlener opened the pull request:
BEGINRELEASENOTES
- Make sure to properly clean up collection names when they are extracted from the Gaudi TES

ENDRELEASENOTES

Fixes #235 

- [ ] Tests
- [ ] Needs key4hep/k4FWCore#312
- [x] Needs #237 
- [x] Includes #239
- [x] Includes #238 
- [x] Includes #243 

@Zehvogel can you give this a go to see if it doesn't break other things?

tmadlener committed:
Always remove leading / to hvae valid coll name

Zehvogel commented:
Lol, now it gets further than before, but still fails in the same line, this time hitting this exception:
https://github.com/iLCSoft/LCIO/blob/96e6a070780b94237ac7728c96b05fd664edd128/src/cpp/src/UTIL/PIDHandler.cc#L102-L103
I added a debug message to check and it still chokes on the `BuildUpVertices` collection, now correctly spelt...
So I cannot say yet if this PR will break anything :D

tmadlener committed:
Don't use potentially unsed id to lookup collection names

tmadlener commented:
I have switched the name lookup away from requiring the collection ID, so I *think* we should no longer mix up collections during the conversion. I didn't fully test this yet, but it should in principle fix the issue.

@Zehvogel can you run some quick checks with this? Or alternatively let me know which branch of CLDConfig to use to test myself.

Zehvogel commented:
@tmadlener you can check it with the `MLJetTagger` test in https://github.com/key4hep/CLDConfig/pull/78

I can probably take a look this afternoon, but maybe you are faster ;)

Zehvogel commented:
Now I am suspicious about:
```
[ctest] ToolSvc.EventNu...  DEBUG Event: 0 Run: 0
[ctest] EventNumber         FATAL Standard std::exception is caught in sysExecute
[ctest] EventNumber         ERROR lcio::DataNotAvailableException: LCEventImpl::getCollection: collection not in event:VertexJets_PID_yth
```
In the LCIO world, `VertexJets_PID_yth` should not be a collection, so this gets wrongly picked up somewhere...

tmadlener commented:
Hmm... Looks like some of the ParticleID collections now escape their enclosure somehow. Let me investigate why.

tmadlener committed:
Fix missing newline in debug outputs

tmadlener committed:
Revert back to using the collection id for lookup

tmadlener commented:
I have switched back to using the collection ID to identify collections to which ParticleID information should be attached. This means that we need the changes in key4hep/k4FWCore#312, especially since we make direct use of one of the newly introduced methods there.

With the latest changes this also switches the storage of converted EDM4hep collections (from EDM4hep) to the TES over from the (to be deprecated) `DataHandle` to use `AnyDataWrapper`, which immediately makes it compatible with Functional algorithms and the IOSvc.

Zehvogel commented:
Probably, we want to make that distinction already earlier, were we just add all the collections to be converted here
https://github.com/tmadlener/k4MarlinWrapper/blob/20386aaa9b64b4cc8ad7e55e0b40f692ca81f57f/k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp#L397-L420

tmadlener committed:
Add more debug output

tmadlener committed:
Use service to look up corresponding LCIO collections

tmadlener commented:
I seem to have solved all issues locally to make https://github.com/key4hep/CLDConfig/pull/78 work (with some minor adaptions to that). I will have to check what I broke in the process and looking at the `convertCollections` function that came out of this, I also think some refactoring of that is still in order.

tmadlener committed:
Do not store in an IOSvc compatible format into TES

Easier to keep collectionID assignment

tmadlener committed:
Fix formatting of warning message

tmadlener commented on k4MarlinWrapper/src/components/Lcio2EDM4hep.cpp:
```diff
@@ -92,9 +92,15 @@ void Lcio2EDM4hepTool::registerCollection(
+  const auto collId = idTable.add(name);
+  e4hColl->setID(collId);
+  debug() << fmt::format("Adding collection '{}' with collection id: {:0>8x}", name, e4hColl->getID()) << endmsg;
+  auto wrapper = new AnyDataWrapper<std::unique_ptr<podio::CollectionBase>>(std::move(e4hColl));
```
Reverted to original version which takes care of the collectionID assignment automatically for now. This will create a bit more work downstream when we remove PodioDataSvc support, but that needs work in k4FWCore first.

tmadlener committed:
Remove commented line

tmadlener committed:
Refactor particle id meta data conversion

tmadlener committed:
Remove unnecessary track keeping of id:name mapping

jmcarcell committed:
Fix leftovers from rebasing onto main

Keep the convertParticleIDs interface and registerCollection in the
modern AnyDataWrapper-based form that main uses, dropping references to
the removed attacheParticleIDMetaInfo helper and the deprecated
DataWrapper approach.
